src/contract.rs

```diff
@@ -14,63 +17,114 @@ pub struct InitMsg {
+#[derive(Serialize, Deserialize, Clone, Debug)]
 pub struct Tally {
     yes: u64,
     no: u64,
 }
 
 pub fn handle<S: Storage, A: Api, Q: Querier>(
     deps: &mut Extern<S, A, Q>,
-    _env: Env,
+    env: Env,
     msg: HandleMsg,
 ) -> HandleResult {
-    let mut tally: Tally = bincode2::deserialize(&deps.storage.get(b"tally").unwrap()).unwrap();
+    let poll_running: bool = deserialize(&deps.storage.get(b"running").unwrap())?;
+    if !poll_running {
+        return Err(StdError::generic_err("the poll is closed"));
+    }
 
-    if msg.yes {
-        tally.yes += 1;
-    } else {
-        tally.no += 1;
+    match msg {
+        HandleMsg::Vote { yes } => {
+            let sender_address = deps.api.canonical_address(&env.message.sender)?;
+
+            let mut storage = PrefixedStorage::new(b"voters", &mut deps.storage);
+            let mut storage = AppendStoreMut::<CanonicalAddr, _>::attach(&mut storage).unwrap()?;
+            if storage
+                .iter()
+                .any(|address| address.unwrap() == sender_address)
+            {
+                return Err(StdError::generic_err("This account has already voted!"));
+            }
+            storage.push(&sender_address)?;
+
+            let mut tally: Tally = deserialize(&deps.storage.get(b"tally").unwrap())?;
+            if yes {
+                tally.yes += 1;
+            } else {
+                tally.no += 1;
+            }
+            deps.storage.set(b"tally", &serialize(&tally)?);
+        }
+        HandleMsg::Close {} => {
+            let admin: HumanAddr = deserialize(&deps.storage.get(b"admin").unwrap())?;
+            if env.message.sender != admin {
+                return Err(StdError::generic_err("only the admin can close the vote"));
+            }
+            deps.storage.set(b"running", &serialize(&false)?)
+        }
     }
 
-    deps.storage
-        .set(b"tally", &bincode2::serialize(&tally).unwrap());
     Ok(HandleResponse::default())
 }
 
 #[derive(Serialize, Deserialize, Clone, PartialEq, JsonSchema)]
 #[serde(rename_all = "snake_case")]
 pub enum QueryMsg {
+    /// Get poll name
     GetPoll {},
+    /// Get the number of votes for/against
     GetTally {},
 }
 
 pub fn query<S: Storage, A: Api, Q: Querier>(deps: &Extern<S, A, Q>, msg: QueryMsg) -> QueryResult {
     match msg {
         QueryMsg::GetPoll {} => {
-            let poll: String = bincode2::deserialize(&deps.storage.get(b"poll").unwrap()).unwrap();
-            Ok(Binary(serde_json_wasm::to_vec(&poll).unwrap()))
+            let poll: String = deserialize(&deps.storage.get(b"poll").unwrap())?;
+            Ok(to_binary(&poll)?)
         }
         QueryMsg::GetTally {} => {
-            let tally: Tally = bincode2::deserialize(&deps.storage.get(b"tally").unwrap()).unwrap();
-            Ok(Binary(serde_json_wasm::to_vec(&tally).unwrap()))
+            let poll_running: bool = deserialize(&deps.storage.get(b"running").unwrap())?;
```

Comment:
In addition to actively closing the poll, I'd add an automatic block height deadline. It can be an opener for talking about this and other stuff you can find under `env`

Reply:
I do use `env.message.sender` a couple times. do you think it's worth the added complexity?

Reply:
imo it's simple enough and it makes more sense in general

Reply:
sure, i'll add it

Reply:
done in 
e7e99b8